### ISSUE TITLE
hero:unfreeze() should renamed to hero:start_free()

### ISSUE THREAD

**Issue by christopho** [solarus-games/solarus#1086]
For consistency with the rest of the API and because it goes back to the "free" state no matter what the current state was (not only when the hero was frozen).

**Comment by christopho** [comment]
Actually hero:unfreeze() keep the carried objet or the loaded sword if any. Maybe we need two functions.

**Comment by Diarandor** [comment]
I think hero:set_free(boolean) would be even better (where the boolean parameter would be optional).

Maybe we should have 2 functions: hero.set_free and hero.restart, where only the second one restarts state, animations, current frame of animation, etc.

**Comment by vlag** [comment]
hero:set_state(state[, keep_actions]) ?

**Comment by Diarandor** [comment]
@vlag: which "actions" should be kept? or what do you mean by actions?

I am not sure if that will be always possible, because some things could be impossible to keep for certain states.

@christopho: I think we should allow somehow to freeze/unfreeze the hero without restarting everything (state, animations, frame, carried entities, timers, etc). Some optional parameter in the function could do it.

**Comment by vlag** [comment]
@Diarandor I'm not really aware about details but I guess the same ones that are currently kept by hero:unfreeze(), or even better, all running actions that are possible to keep with the new state.
The only other option I see with this kind of function is to use the parameter as a table and precise what built-in actions should be kept if possible, but I tend to prefer the boolean way.

By actions I meant what Christopho said before, so keep the carried objet or the sword loaded for instance.
And yes of course sometimes keeping an action with a certain state won't be possible, but this is the responsability of the quest maker so I guess trigger a warning in this case is enough.

**Comment by Diarandor** [comment]
Ok, so I guess that by state you only refer to the "free"/"frozen" states, and not others like pushing, pulling, swimming, etc. In that case, instead of "hero:set_state(state[, keep_actions])", I propose the clearer syntax: hero:set_frozen(boolean [, keep_actions]).

**Comment by vlag** [comment]
If the free and frozen states would be the only ones that allow actions in the same time, then you would be right but I think there are some others, like jumping and holding the sword for instance (don't know if it's possible yet), or forced walking and carrying an object.

**Comment by Diarandor** [comment]
-Using the built-in sword while jumping is possible with a custom jump, and that is already done. (And the custom jump is much better than the current built-in one since it allows to control the hero during the jump.)
-I completely agree with you that we should allow to carry objects and start other actions at the same time. Maybe it will be possible to script that when Solarus 1.6 adds some functions like hero:get_carried_entity() and hero:set_carried_entity(entity), but I am not sure.

**Comment by vlag** [comment]
Yes indeed but actually my point was that hero:set_state(state[, keep_actions]) may cover more use cases than hero:set_frozen(boolean [, keep_actions]), so it's not only a question of syntax to me :)

**Comment by Diarandor** [comment]
Ok, then I agree. That would reduce the number of functions of the Lua API, and even allows to extend that function to custom states (from the scripts). However, I think we would need an extra parameter:
hero:set_state(state, value [, properties])
state: a string. value: boolean. properties: a list.
